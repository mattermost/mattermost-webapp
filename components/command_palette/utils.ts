// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Channel} from '@mattermost/types/channels';
import {Team} from '@mattermost/types/teams';

// See LICENSE.txt for license information.
import {CommandPaletteItem} from './command_palette_list_item/command_palette_list_item';
import {
    CmdPalettePictographIcon,
    CmdPalettePictographType,
} from './command_palette_list_item/command_palette_list_item_pictograph';
import {ChannelType, CommandPaletteEntities} from './types';

const channelTypeOrder = {
    [ChannelType.OPEN_CHANNEL]: 0,
    [ChannelType.PRIVATE_CHANNEL]: 1,
    [ChannelType.DM_CHANNEL]: 2,
    [ChannelType.GM_CHANNEL]: 3,
};

function filterName(name: string): string {
    return name.replace(/[.,'"\/#!$%\^&\*;:{}=\-_`~()]/g, ''); // eslint-disable-line no-useless-escape
}

export function channelToCommandPaletteItemTransformer(channels: Channel[], teams: Record<string, Team>): CommandPaletteItem[] {
    return channels.map((channel) => {
        const isArchived = channel.delete_at ? channel.delete_at !== 0 : false;

        let channelIcon = CmdPalettePictographIcon.GLOBE;
        if (isArchived) {
            channelIcon = CmdPalettePictographIcon.ARCHIVE;
        }
        return {
            description: channel.type === ChannelType.DM_CHANNEL ? channel.name : '',
            id: channel.id,
            isDeactivated: isArchived,
            teamName: teams[channel.team_id]?.name,
            pictograph: {
                type: CmdPalettePictographType.ICON,
                pictographItem: channelIcon,
            },
            subType: channel.type as ChannelType,
            title: channel.display_name,
            type: CommandPaletteEntities.Channel,
            isArchived,
        };
    });
}

function sortChannelsByTypeAndDisplayName(locale: string, cmdPaletteItemA: CommandPaletteItem, cmdPaletteItemB: CommandPaletteItem): number {
    if (channelTypeOrder[cmdPaletteItemA.subType] !== channelTypeOrder[cmdPaletteItemB.subType]) {
        if (channelTypeOrder[cmdPaletteItemA.subType] < channelTypeOrder[cmdPaletteItemB.subType]) {
            return -1;
        }

        return 1;
    }

    const aDisplayName = filterName(cmdPaletteItemA.title);
    const bDisplayName = filterName(cmdPaletteItemB.title);

    if (aDisplayName !== bDisplayName) {
        return aDisplayName.toLowerCase().localeCompare(bDisplayName.toLowerCase(), locale, {numeric: true});
    }

    return cmdPaletteItemA.title.toLowerCase().localeCompare(cmdPaletteItemA.title.toLowerCase(), locale, {numeric: true});
}

function sortChannelsByRecencyAndTypeAndDisplayName(cmdPaletteItemA: CommandPaletteItem, cmdPaletteItemB: CommandPaletteItem) {
    if (cmdPaletteItemA.lastViewedAt && cmdPaletteItemB.lastViewedAt) {
        return cmdPaletteItemB.lastViewedAt - cmdPaletteItemA.lastViewedAt;
    } else if (cmdPaletteItemA.lastViewedAt) {
        return -1;
    } else if (cmdPaletteItemB.lastViewedAt) {
        return 1;
    }

    // MM-12677 When this is migrated this needs to be fixed to pull the user's locale
    return sortChannelsByTypeAndDisplayName('en', cmdPaletteItemA, cmdPaletteItemB);
}

export function commandPaletteSorter(prefix: string, cmdPaletteItemA: CommandPaletteItem, cmdPaletteItemB: CommandPaletteItem) {
    if (cmdPaletteItemA.isArchived && !cmdPaletteItemB.isArchived) {
        return 1;
    } else if (!cmdPaletteItemA.isArchived && cmdPaletteItemB.isArchived) {
        return -1;
    }

    if (cmdPaletteItemA.isDeactivated && !cmdPaletteItemB.isDeactivated) {
        return 1;
    } else if (cmdPaletteItemB.isDeactivated && !cmdPaletteItemA.isDeactivated) {
        return -1;
    }

    let aDisplayName = cmdPaletteItemA.title.toLowerCase();
    let bDisplayName = cmdPaletteItemA.title.toLowerCase();

    if (cmdPaletteItemA.type === CommandPaletteEntities.Channel &&
        cmdPaletteItemA.subType === ChannelType.DM_CHANNEL &&
        aDisplayName.startsWith('@')) {
        aDisplayName = aDisplayName.substring(1);
    }

    if (cmdPaletteItemB.type === CommandPaletteEntities.Channel &&
        cmdPaletteItemB.subType === ChannelType.DM_CHANNEL && bDisplayName.startsWith('@')) {
        bDisplayName = bDisplayName.substring(1);
    }

    const aStartsWith = aDisplayName.startsWith(prefix) || cmdPaletteItemA.description.toLowerCase().startsWith(prefix);
    const bStartsWith = bDisplayName.startsWith(prefix) || cmdPaletteItemB.description.toLowerCase().startsWith(prefix);

    // Open channels user haven't interacted should be at the  bottom of the list
    if (cmdPaletteItemA.type === CommandPaletteEntities.Channel &&
        cmdPaletteItemA.subType === ChannelType.OPEN_CHANNEL &&
        !cmdPaletteItemA.lastViewedAt &&
        ((cmdPaletteItemB.type === CommandPaletteEntities.Channel &&
            cmdPaletteItemB.subType !== ChannelType.OPEN_CHANNEL) || cmdPaletteItemB.lastViewedAt)) {
        return 1;
    } else if ((cmdPaletteItemB.type === CommandPaletteEntities.Channel &&
        cmdPaletteItemB.subType === ChannelType.OPEN_CHANNEL) || !cmdPaletteItemB.lastViewedAt) {
        return -1;
    }

    // Sort channels starting with the search term first
    if (aStartsWith && !bStartsWith) {
        return -1;
    } else if (!aStartsWith && bStartsWith) {
        return 1;
    }
    return sortChannelsByRecencyAndTypeAndDisplayName(cmdPaletteItemA, cmdPaletteItemB);
}
