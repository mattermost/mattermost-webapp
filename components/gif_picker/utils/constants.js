// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export default {
    ItemTapAction: {
        OPEN_EMBED_PAGE: 1,
        SHARE: 2,
    },

    Tab: {
        TRENDING: 0,
        REACTIONS: 1,
    },

    appName: {
        mattermost: 'mattermost',
    },
};
