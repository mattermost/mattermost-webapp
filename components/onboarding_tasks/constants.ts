// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

const FINISHED = 999;

export const OnboardingTaskCategory = 'onboarding_task_list';

// Whole task list is based on these
export const OnboardingTasksName = {
    CREATE_FROM_WORK_TEMPLATE: 'create_from_work_template',
    CHANNELS_TOUR: 'channels_tour',
    BOARDS_TOUR: 'boards_tour',
    PLAYBOOKS_TOUR: 'playbooks_tour',
    INVITE_PEOPLE: 'invite_people',
    DOWNLOAD_APP: 'download_app',
    COMPLETE_YOUR_PROFILE: 'complete_your_profile',
    EXPLORE_OTHER_TOOLS: 'explore_other_tools',
    VISIT_SYSTEM_CONSOLE: 'visit_system_console',
    START_TRIAL: 'start_trial',
};

export const OnboardingTaskList = {
    ONBOARDING_TASK_LIST_OPEN: 'onboarding_task_list_open',
    ONBOARDING_TASK_LIST_SHOW: 'onboarding_task_list_show',
    ONBOARDING_TASK_LIST_CLOSE: 'onboarding_task_list_close',
    ONBOARDING_VIDEO_MODAL: 'onboarding_video_modal',
    DECLINED_ONBOARDING_TASK_LIST: 'declined_onboarding_task_list',

    // auto show channel linked boards A/B test
    ONBOARDING_LINKED_BOARD_AUTO_SHOWN: 'linked_board_auto_shown',
};

export const GenericTaskSteps = {
    START: 0,
    STARTED: 1,
    FINISHED,
};

export const TaskNameMapToSteps = {
    [OnboardingTasksName.CREATE_FROM_WORK_TEMPLATE]: GenericTaskSteps,
    [OnboardingTasksName.CHANNELS_TOUR]: GenericTaskSteps,
    [OnboardingTasksName.BOARDS_TOUR]: GenericTaskSteps,
    [OnboardingTasksName.PLAYBOOKS_TOUR]: GenericTaskSteps,
    [OnboardingTasksName.COMPLETE_YOUR_PROFILE]: GenericTaskSteps,
    [OnboardingTasksName.EXPLORE_OTHER_TOOLS]: GenericTaskSteps,
    [OnboardingTasksName.DOWNLOAD_APP]: GenericTaskSteps,
    [OnboardingTasksName.VISIT_SYSTEM_CONSOLE]: GenericTaskSteps,
    [OnboardingTasksName.INVITE_PEOPLE]: GenericTaskSteps,
    [OnboardingTasksName.START_TRIAL]: GenericTaskSteps,
};

