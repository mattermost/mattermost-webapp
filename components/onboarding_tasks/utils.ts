// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export const generateTelemetryTag = (category: string, name: string, suffix: string) => {
    return `${category}--${name}--${suffix}`;
};

export function getAnalyticsCategory(isAdmin: boolean) {
    return isAdmin ? 'admin_onboarding' : 'end_user_onboarding';
}
