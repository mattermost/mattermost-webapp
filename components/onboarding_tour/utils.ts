// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {FINISHED} from './constant';

export const KeyCodes: Record<string, [string, number]> = {
    ENTER: ['Enter', 13],
    COMPOSING: ['Composing', 229],
};

// This is extracted from utils file to remove dependency on utils file of webapp
export function isKeyPressed(event: KeyboardEvent, key: [string, number]): boolean {
    // There are two types of keyboards
    // 1. English with different layouts(Ex: Dvorak)
    // 2. Different language keyboards(Ex: Russian)

    if (event.keyCode === KeyCodes.COMPOSING[1]) {
        return false;
    }

    // checks for event.key for older browsers and also for the case of different English layout keyboards.
    if (typeof event.key !== 'undefined' && event.key !== 'Unidentified' && event.key !== 'Dead') {
        const isPressedByCode = event.key === key[0] || event.key === key[0].toUpperCase();
        if (isPressedByCode) {
            return true;
        }
    }

    // used for different language keyboards to detect the position of keys
    return event.keyCode === key[1];
}

export const getLastStep = (Steps: Record<string, number>) => {
    return Object.values(Steps).reduce((maxStep, candidateMaxStep) => {
        // ignore the "opt out" FINISHED step as the max step.
        if (candidateMaxStep > maxStep && candidateMaxStep !== FINISHED) {
            return candidateMaxStep;
        }
        return maxStep;
    }, Number.MIN_SAFE_INTEGER);
};
