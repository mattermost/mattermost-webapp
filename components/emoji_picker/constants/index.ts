// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {EmojiCategory} from '@mattermost/types/emojis';

import * as Emoji from 'utils/emoji';

import {Category, Categories} from '../types';

export const RECENT = 'recent';
export const SEARCH_RESULTS = 'searchResults';
export const SMILEY_EMOTION = 'smileys-emotion';
export const CUSTOM = 'custom';

const categoryClass: Map<EmojiCategory, string> = new Map([
    [RECENT, 'icon-clock-outline'],
    [SMILEY_EMOTION, 'icon-emoticon-happy-outline'],
    ['people-body', 'icon-account-outline'],
    ['animals-nature', 'icon-leaf-outline'],
    ['food-drink', 'icon-food-apple'],
    ['activities', 'icon-basketball'],
    ['travel-places', 'icon-airplane-variant'],
    ['objects', 'icon-lightbulb-outline'],
    ['symbols', 'icon-heart-outline'],
    ['flags', 'icon-flag-outline'],
    [CUSTOM, 'icon-emoticon-custom-outline'],
    [SEARCH_RESULTS, ''],
]);

function createCategory(name: EmojiCategory): Category {
    return {
        name,
        id: Emoji.CategoryTranslations.get(name) || '',
        className: categoryClass.get(name) || '',
        message: Emoji.CategoryMessage.get(name)!,
    };
}

export const RECENT_EMOJI_CATEGORY: Pick<Categories, 'recent'> = {recent: createCategory(RECENT)};
export const SEARCH_EMOJI_CATEGORY: Pick<Categories, typeof SEARCH_RESULTS> = {searchResults: createCategory(SEARCH_RESULTS)};

export const CATEGORIES: Categories = Emoji.CategoryNames.
    filter((category) => !(category === 'recent' || category === 'searchResults')).
    reduce((previousCategory, currentCategory) => {
        return {
            ...previousCategory,
            [currentCategory]: createCategory(currentCategory as EmojiCategory),
        };
    }, {} as Categories);

export const EMOJI_PER_ROW = 9; // needs to match variable `$emoji-per-row` in _variables.scss
export const ITEM_HEIGHT = 36; //as per .emoji-picker__item height in _emoticons.scss
export const EMOJI_CONTAINER_HEIGHT = 290; // If this changes, the spaceRequiredAbove and spaceRequiredBelow props passed to the EmojiPickerOverlay must be updated

export const CATEGORY_HEADER_ROW = 'categoryHeaderRow';
export const EMOJIS_ROW = 'emojisRow';

export const EMOJI_SCROLL_THROTTLE_DELAY = 150;
export const EMOJI_ROWS_OVERSCAN_COUNT = 1;

export const CUSTOM_EMOJIS_PER_PAGE = 200;
export const CUSTOM_EMOJI_SEARCH_THROTTLE_TIME_MS = 1000;
