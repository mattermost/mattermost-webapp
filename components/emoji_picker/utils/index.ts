// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Emoji, EmojiCategory} from 'mattermost-redux/types/emojis';

import {compareEmojis} from 'utils/emoji_utils';

import {Categories} from '../types';

function sortEmojis(emojis: Emoji[], recentEmojiStrings: string[], filter: string) {
    const recentEmojis: Emoji[] = [];
    const emojisMinusRecent: Emoji[] = [];

    Object.values(emojis).forEach((emoji) => {
        let emojiArray = emojisMinusRecent;
        const alias = 'short_names' in emoji ? emoji.short_names : [emoji.name];
        for (let i = 0; i < alias.length; i++) {
            if (recentEmojiStrings.includes(alias[i].toLowerCase())) {
                emojiArray = recentEmojis;
            }
        }

        emojiArray.push(emoji);
    });

    return [
        ...recentEmojis.sort((firstEmoji, secondEmoji) => compareEmojis(firstEmoji, secondEmoji, filter)),
        ...emojisMinusRecent.sort((firstEmoji, secondEmoji) => compareEmojis(firstEmoji, secondEmoji, filter)),
    ];
}

function getFilteredEmojis(allEmojis: Record<string, Emoji>, filter: string, recentEmojisString: string[]): Emoji[] {
    const emojis = Object.values(allEmojis).filter((emoji) => {
        const alias = 'short_names' in emoji ? emoji.short_names : [emoji.name];
        for (let i = 0; i < alias.length; i++) {
            if (alias[i].toLowerCase().includes(filter)) {
                return true;
            }
        }

        return false;
    });

    return sortEmojis(emojis, recentEmojisString, filter);
}

export function getEmojisByCategory(allEmojis: Record<string, Emoji>, categories: Categories, categoryName: EmojiCategory, filter: string, recentEmojisString: string[]): Emoji[] {
    if (filter.length) {
        return getFilteredEmojis(allEmojis, filter, recentEmojisString);
    }

    // if (categoryName === 'recent') {
    //     const recentEmojiIds = this.state.categories?.recent?.emojiIds ?? [];
    //     if (recentEmojiIds.length === 0) {
    //         return [];
    //     }

    //     const recentEmojis = new Map();

    //     recentEmojiIds.forEach((emojiId) => {
    //         const emoji = this.state.allEmojis[emojiId];
    //         const emojiSkin = getSkin(emoji);

    //         if (emojiSkin) {
    //             const emojiWithUserSkin = this.convertEmojiToUserSkinTone(emoji, emojiSkin, this.props.userSkinTone);
    //             recentEmojis.set(emojiWithUserSkin.unified, emojiWithUserSkin);
    //         } else {
    //             recentEmojis.set(emojiId, emoji);
    //         }
    //     });

    //     return Array.from(recentEmojis.values());
    // }

    return categories[categoryName].emojiIds.map((emojiId) => allEmojis[emojiId]);
}
