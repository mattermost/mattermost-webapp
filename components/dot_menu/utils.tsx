// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

import {trackEvent} from 'actions/telemetry_actions';

import {EventTypes, TELEMETRY_CATEGORIES} from 'utils/constants';

export type ChangeEvent = React.KeyboardEvent | React.MouseEvent;

export const trackDotMenuEvent = (e: ChangeEvent, suffix: string): void => {
    if (e.type === EventTypes.CLICK) {
        trackEvent(TELEMETRY_CATEGORIES.POST_INFO_MORE, EventTypes.CLICK + '_' + suffix);
    } else {
        trackEvent(TELEMETRY_CATEGORIES.POST_INFO_MORE, EventTypes.SHORTCUT + '_ ' + suffix);
    }
};
