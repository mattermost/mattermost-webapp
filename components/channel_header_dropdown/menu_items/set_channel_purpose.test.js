// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';
import {Permissions} from 'mattermost-redux/constants';

import {Constants} from 'utils/constants';

import SetChannelPurpose from './set_channel_purpose';

describe('components/ChannelHeaderDropdown/MenuItem.SetChannelPurpose', () => {
    const baseProps = {
        channel: {
            id: 'channel_id',
            team_id: 'team_id',
            type: Constants.OPEN_CHANNEL,
        },
        isArchived: false,
        isReadonly: false,
    };

    it('should match snapshot', () => {
        const wrapper = shallow(<SetChannelPurpose {...baseProps}/>);

        expect(wrapper).toMatchSnapshot();
    });

    it('should be hidden if the channel is archived', () => {
        const props = {
            ...baseProps,
            isArchived: true,
        };
        const wrapper = shallow(<SetChannelPurpose {...props}/>);

        expect(wrapper.isEmptyRender()).toBeTruthy();
    });

    it('should be hidden if the channel is readonly', () => {
        const props = {
            ...baseProps,
            isReadonly: true,
        };
        const wrapper = shallow(<SetChannelPurpose {...props}/>);

        expect(wrapper.isEmptyRender()).toBeTruthy();
    });

    it('should be hidden if the channel type is DM or GM', () => {
        const props = {
            ...baseProps,
            channel: {...baseProps.channel},
        };
        const makeWrapper = () => shallow(<SetChannelPurpose {...props}/>);

        props.channel.type = Constants.DM_CHANNEL;
        expect(makeWrapper().isEmptyRender()).toBeTruthy();

        props.channel.type = Constants.GM_CHANNEL;
        expect(makeWrapper().isEmptyRender()).toBeTruthy();
    });

    it('should requires right permission level for channel type to manage purpose', () => {
        const props = {
            ...baseProps,
            channel: {...baseProps.channel},
        };
        const makeWrapper = () => shallow(<SetChannelPurpose {...props}/>);

        props.channel.type = Constants.OPEN_CHANNEL;
        expect(makeWrapper().prop('permissions')[0]).toBe(Permissions.MANAGE_PUBLIC_CHANNEL_PROPERTIES);

        props.channel.type = Constants.PRIVATE_CHANNEL;
        expect(makeWrapper().prop('permissions')[0]).toBe(Permissions.MANAGE_PRIVATE_CHANNEL_PROPERTIES);
    });
});
