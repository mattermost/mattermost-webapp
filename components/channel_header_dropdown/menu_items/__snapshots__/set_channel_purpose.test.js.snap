// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ChannelHeaderDropdown/MenuItem.SetChannelPurpose should match snapshot 1`] = `
<Connect(ChannelPermissionGate)
  channelId="channel_id"
  permissions={
    Array [
      "manage_public_channel_properties",
    ]
  }
  teamId="team_id"
>
  <li
    role="presentation"
  >
    <Connect(ModalToggleButtonRedux)
      dialogProps={
        Object {
          "channel": Object {
            "id": "channel_id",
            "team_id": "team_id",
            "type": "O",
          },
        }
      }
      dialogType={[Function]}
      modalId="edit_channel_purpose"
      role="menuitem"
    >
      <FormattedMessage
        defaultMessage="Edit Channel Purpose"
        id="channel_header.setPurpose"
        values={Object {}}
      />
    </Connect(ModalToggleButtonRedux)>
  </li>
</Connect(ChannelPermissionGate)>
`;
